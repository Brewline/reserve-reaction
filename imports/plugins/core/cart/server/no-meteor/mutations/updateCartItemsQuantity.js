import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import { Cart as CartSchema } from "/imports/collections/schemas";
import getCartById from "../util/getCartById";

const inputSchema = new SimpleSchema({
  "cartId": String,
  "items": {
    type: Array,
    minCount: 1
  },
  "items.$": Object,
  "items.$.cartItemId": String,
  "items.$.quantity": {
    type: SimpleSchema.Integer,
    min: 0
  },
  "token": {
    type: String,
    optional: true
  }
});

/**
 * @method updateCartItemsQuantity
 * @summary Sets a new quantity for one or more items in a cart
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - Necessary input
 * @param {String} input.cartId - The ID of the cart in which all of the items exist
 * @param {String} input.items - Array of items to update
 * @param {Number} input.items.cartItemId - The cart item ID
 * @param {Object} input.items.quantity - The new quantity, which must be an integer of 0 or greater
 * @param {String} input.token - The token if the cart is an anonymous cart
 * @return {Promise<Object>} An object containing the updated cart in a `cart` property
 */
export default async function updateCartItemsQuantity(context, input) {
  inputSchema.validate(input || {});

  const { cartId, items, token: cartToken } = input;

  const cart = await getCartById(context, cartId, { cartToken, throwIfNotFound: true });

  const updatedItems = cart.items.reduce((list, item) => {
    const update = items.find(({ cartItemId }) => cartItemId === item._id);
    if (!update) {
      list.push({ ...item });
    } else if (update.quantity > 0) {
      // Update quantity as instructed, while omitting the item if quantity is 0
      list.push({
        ...item,
        quantity: update.quantity,
        // Update the subtotal since it is a multiple of the price
        subtotal: {
          amount: item.price.amount * update.quantity,
          currencyCode: item.subtotal.currencyCode
        }
      });
    }
    return list;
  }, []);

  const { appEvents, collections } = context;
  const { Cart } = collections;

  const updatedAt = new Date();
  const modifier = {
    $set: {
      items: updatedItems,
      updatedAt
    }
  };
  CartSchema.validate(modifier, { modifier: true });

  const { matchedCount } = await Cart.updateOne({ _id: cartId }, modifier);
  if (matchedCount === 0) throw new ReactionError("server-error", "Failed to update cart");

  const updatedCart = { ...cart, items: updatedItems, updatedAt };

  await appEvents.emit("afterCartUpdate", updatedCart);

  return { cart: updatedCart };
}
