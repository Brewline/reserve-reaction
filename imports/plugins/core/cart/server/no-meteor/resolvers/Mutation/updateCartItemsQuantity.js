import { decodeCartItemOpaqueId, decodeCartOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/cart";

/**
 * @name "Mutation.updateCartItemsQuantity"
 * @method
 * @memberof Cart/GraphQL
 * @summary resolver for the updateCartItemsQuantity GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.cartId - The ID of the cart in which all of the items exist
 * @param {String} args.input.items - Array of items to update
 * @param {Number} args.input.items.cartItemId - The cart item ID
 * @param {Object} args.input.items.quantity - The new quantity, which must be an integer of 0 or greater
 * @param {String} args.input.token - The token if the cart is an anonymous cart
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} UpdateCartItemsQuantityPayload
 */
export default async function updateCartItemsQuantity(parentResult, { input }, context) {
  const { cartId: opaqueCartId, clientMutationId = null, items: itemsInput, token } = input;

  const cartId = decodeCartOpaqueId(opaqueCartId);
  const items = itemsInput.map((item) => ({ cartItemId: decodeCartItemOpaqueId(item.cartItemId), quantity: item.quantity }));

  const { cart } = await context.mutations.updateCartItemsQuantity(context, {
    cartId,
    items,
    token
  });

  return { cart, clientMutationId };
}
