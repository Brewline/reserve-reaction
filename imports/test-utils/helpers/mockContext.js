const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
