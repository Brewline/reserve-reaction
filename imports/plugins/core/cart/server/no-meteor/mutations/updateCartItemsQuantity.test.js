import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import updateCartItemsQuantity from "./updateCartItemsQuantity";


const cartItemId1 = Factory.CartItem.makeOne({
  _id: "cartItemId1",
  quantity: 5,
  price: {
    amount: 400,
    currencyCode: "mockCurrencyCode"
  },
  subtotal: {
    amount: 2000,
    currencyCode: "mockCurrencyCode"
  }
});
const cartItemId2 = Factory.CartItem.makeOne({
  _id: "cartItemId2",
  quantity: 5,
  price: {
    amount: 200,
    currencyCode: "mockCurrencyCode"
  },
  subtotal: {
    amount: 1000,
    currencyCode: "mockCurrencyCode"
  }
});

const dbCart = {
  _id: "cartId",
  items: [
    cartItemId1, cartItemId2
  ]
};

const updatedCartItemId1 = Factory.CartItem.makeOne({
  subtotal: {
    amount: 2000,
    currencyCode: "mockCurrencyCode"
  },
  ...cartItemId1
});
const updatedCartItemId2 = Factory.CartItem.makeOne({
  subtotal: {
    amount: 1000,
    currencyCode: "mockCurrencyCode"
  },
  ...cartItemId2
});

const updatedDbCart = {
  _id: "cartId",
  items: [
    updatedCartItemId1, updatedCartItemId2
  ]
};

const updatedDbCartAfterRemoval = {
  _id: "cartId",
  items: [
    updatedCartItemId2
  ]
};

beforeEach(() => {
  jest.clearAllMocks();
});

test("updates the quantity of multiple items in account cart", async () => {
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(dbCart));
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(updatedDbCart));

  const result = await updateCartItemsQuantity(mockContext, {
    cartId: "cartId",
    items: [
      {
        cartItemId: "cartItemId1",
        quantity: 1
      },
      {
        cartItemId: "cartItemId2",
        quantity: 2
      }
    ]
  });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    _id: "cartId"
  }, {
    $set: {
      items: [
        {
          ...dbCart.items[0],
          quantity: 1,
          subtotal: {
            amount: dbCart.items[0].price.amount,
            currencyCode: dbCart.items[0].subtotal.currencyCode
          }
        },
        {
          ...dbCart.items[1],
          quantity: 2,
          subtotal: {
            amount: dbCart.items[1].price.amount * 2,
            currencyCode: dbCart.items[1].subtotal.currencyCode
          }
        }
      ],
      updatedAt: jasmine.any(Date)
    }
  });

  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({
    _id: "cartId"
  });

  expect(result).toEqual({
    cart: {
      _id: "cartId",
      items: [
        {
          ...updatedDbCart.items[0],
          quantity: 5,
          subtotal: {
            amount: updatedDbCart.items[0].price.amount * 5,
            currencyCode: updatedDbCart.items[0].subtotal.currencyCode
          }
        },
        {
          ...updatedDbCart.items[1],
          quantity: 5,
          subtotal: {
            amount: updatedDbCart.items[1].price.amount * 5,
            currencyCode: updatedDbCart.items[1].subtotal.currencyCode
          }
        }
      ]
    }
  });
});

test("updates the quantity of multiple items in anonymous cart", async () => {
  const hashedToken = "+YED6SF/CZIIVp0pXBsnbxghNIY2wmjIVLsqCG4AN80=";

  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(dbCart));
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(updatedDbCart));

  const cachedAccountId = mockContext.accountId;
  mockContext.accountId = null;
  const result = await updateCartItemsQuantity(mockContext, {
    cartId: "cartId",
    items: [
      {
        cartItemId: "cartItemId1",
        quantity: 1
      },
      {
        cartItemId: "cartItemId2",
        quantity: 2
      }
    ],
    token: "TOKEN"
  });
  mockContext.accountId = cachedAccountId;

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    _id: "cartId"
  }, {
    $set: {
      items: [
        {
          ...dbCart.items[0],
          quantity: 1,
          subtotal: {
            amount: dbCart.items[0].price.amount,
            currencyCode: dbCart.items[0].subtotal.currencyCode
          }
        },
        {
          ...dbCart.items[1],
          quantity: 2,
          subtotal: {
            amount: dbCart.items[1].price.amount * 2,
            currencyCode: dbCart.items[1].subtotal.currencyCode
          }
        }
      ],
      updatedAt: jasmine.any(Date)
    }
  });

  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({
    _id: "cartId",
    anonymousAccessToken: hashedToken
  });

  expect(result).toEqual({
    cart: {
      _id: "cartId",
      items: [
        {
          ...updatedDbCart.items[0],
          quantity: 5,
          subtotal: {
            amount: updatedDbCart.items[0].price.amount * 5,
            currencyCode: updatedDbCart.items[0].subtotal.currencyCode
          }
        },
        {
          ...updatedDbCart.items[1],
          quantity: 5,
          subtotal: {
            amount: updatedDbCart.items[1].price.amount * 5,
            currencyCode: updatedDbCart.items[1].subtotal.currencyCode
          }
        }
      ]
    }
  });
});

test("throws when no account and no token passed", async () => {
  const cachedAccountId = mockContext.accountId;
  mockContext.accountId = null;

  await expect(updateCartItemsQuantity(mockContext, {
    cartId: "cartId",
    items: [
      {
        cartItemId: "cartItemId1",
        quantity: 1
      },
      {
        cartItemId: "cartItemId2",
        quantity: 2
      }
    ]
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = cachedAccountId;
});

test("removes an item if quantity is 0", async () => {
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(dbCart));
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(updatedDbCartAfterRemoval));

  const result = await updateCartItemsQuantity(mockContext, {
    cartId: "cartId",
    items: [
      {
        cartItemId: "cartItemId1",
        quantity: 0
      },
      {
        cartItemId: "cartItemId2",
        quantity: 2
      }
    ]
  });

  expect(mockContext.collections.Cart.updateOne).toHaveBeenCalledWith({
    _id: "cartId"
  }, {
    $set: {
      items: [
        {
          ...dbCart.items[1],
          quantity: 2,
          subtotal: {
            amount: dbCart.items[1].price.amount * 2,
            currencyCode: dbCart.items[1].subtotal.currencyCode
          }
        }
      ],
      updatedAt: jasmine.any(Date)
    }
  });

  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({
    _id: "cartId"
  });

  expect(result).toEqual({
    cart: {
      _id: "cartId",
      items: [
        {
          ...updatedDbCartAfterRemoval.items[0],
          quantity: 5,
          subtotal: {
            amount: updatedDbCartAfterRemoval.items[0].price.amount * 5,
            currencyCode: updatedDbCartAfterRemoval.items[0].subtotal.currencyCode
          }
        }
      ]
    }
  });
});
