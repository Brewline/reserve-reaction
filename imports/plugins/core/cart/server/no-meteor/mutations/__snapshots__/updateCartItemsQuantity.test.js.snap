// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when no account and no token passed 1`] = `"Cart not found"`;
